src/components/ScrollTrack/ScrollTrack.js

```diff
@@ -142,21 +159,66 @@ class ScrollTrack extends Component {
     const { parentWidth, trackWidth } = this.getNodeWidths()
     let nextForward = this.state.left - parentWidth
     const fullForward = parentWidth - trackWidth
+    const { onBeforeNext, onAfterNext } = this.props
 
     // already is, or is going to be, full forward
     if (nextForward <= fullForward) { nextForward = fullForward }
 
-    this.setState({ left: nextForward }, this.computeSlideAttributes)
+    const callbackProps = {
+      atStart: trackWidth <= parentWidth,
+      atEnd: fullForward === nextForward,
+      slideTo: nextForward,
+      parentWidth,
+      trackWidth
+    }
+
+    if (!onBeforeNext) {
+      this.updateLeftValue({
+        left: nextForward,
+        callback: onAfterNext,
+        callbackProps
+      })
+    } else {
+      onBeforeNext(callbackProps).then(() => {
+        this.updateLeftValue({
+          left: nextForward,
+          callback: onAfterNext,
+          callbackProps
+        })
+      })
+    }
   }
 
   slideBack = () => {
-    const { parentWidth } = this.getNodeWidths()
+    const { parentWidth, trackWidth } = this.getNodeWidths()
     let nextBack = this.state.left + parentWidth
+    const { onBeforeBack, onAfterBack } = this.props
 
     // already is, or is going to be, full back
     if (this.state.left >= 0 || nextBack >= 0) { nextBack = 0 }
 
-    this.setState({ left: nextBack }, this.computeSlideAttributes)
+    const callbackProps = {
+      atStart: nextBack === 0,
+      atEnd: false,
+      slideTo: nextBack,
+      parentWidth,
+      trackWidth
+    }
+
+    onBeforeBack && onBeforeBack(callbackProps)
```

Comment:
Total personal preference but I've been liking adding a no-op as a default prop to handle these cases, e.g.

```
ScrollTrack.defaultProps = {
  onBeforeBack: () => {} 
}
```

If we bring in lodash sometime can also do `_.noop`

Reply:
Yeah, I like this. I'll add that in too